File: be_camphub/src/main/java/org/camphub/be_camphub/mapper/DisputeMapper.java
package org.camphub.be_camphub.mapper;

import org.camphub.be_camphub.dto.request.dispute.DisputeCreationRequest;
import org.camphub.be_camphub.dto.response.dispute.DisputeResponse;
import org.camphub.be_camphub.entity.Dispute;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

@Mapper(componentModel = "spring", uses = MediaResourceMapper.class)
public interface DisputeMapper {
    @Mapping(target = "damageTypeName", ignore = true)
    @Mapping(target = "compensationRate", ignore = true)
    @Mapping(target = "reporterName", ignore = true)
    @Mapping(target = "defenderName", ignore = true)
    @Mapping(target = "adminName", ignore = true)
    @Mapping(target = "evidenceUrls", source = "evidences", qualifiedByName = "toResponse")
    DisputeResponse entityToResponse(Dispute dispute);

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "reporterId", ignore = true)
    @Mapping(target = "defenderId", ignore = true)
    @Mapping(target = "status", constant = "PENDING_REVIEW")
    @Mapping(target = "adminDecision", ignore = true)
    @Mapping(target = "compensationAmount", ignore = true)
    @Mapping(target = "adminNote", ignore = true)
    @Mapping(target = "adminId", ignore = true)
    @Mapping(target = "resolvedAt", ignore = true)
    @Mapping(target = "createdAt", expression = "java(java.time.LocalDateTime.now())")
    @Mapping(target = "evidences", source = "evidenceUrls", qualifiedByName = "fromRequest")
    @Mapping(target = "description", source = "note")
    Dispute creationRequestToEntity(DisputeCreationRequest request);
}

File: fe_camphub/src/libs/core/mapping/dispute.mapper.ts
// dispute.mapper.ts
import { createMapper, createMap } from '@automapper/core';
import { pojos, PojosMetadataMap } from '@automapper/pojos';

import { Dispute } from '../types';
import {
    DisputeCreationRequest,
    AdminReviewDisputeRequest
} from '../dto/request';
import { DisputeResponse } from '../dto/response';

// ========================================================
// Mapper
// ========================================================
export const disputeMapper = createMapper({
    strategyInitializer: pojos(),
});

// ========================================================
// Metadata
// ========================================================
PojosMetadataMap.create<Dispute>('Dispute', {
    id: String,
    bookingId: String,
    reporterId: String,
    defenderId: String,
    adminId: String,

    reporterName: String,
    defenderName: String,
    adminName: String,

    description: String,
    evidenceUrls: Array,
    damageTypeId: String,
    damageTypeName: String,
    compensationRate: Number,

    status: String,
    adminDecision: String,
    compensationAmount: Number,
    adminNote: String,

    createdAt: String,
    resolvedAt: String,
});

PojosMetadataMap.create<DisputeResponse>('DisputeResponse', {
    id: String,
    bookingId: String,
    reporterId: String,
    defenderId: String,
    adminId: String,

    reporterName: String,
    defenderName: String,
    adminName: String,

    description: String,
    evidenceUrls: Array,
    damageTypeId: String,
    damageTypeName: String,
    compensationRate: Number,

    status: String,
    adminDecision: String,
    compensationAmount: Number,
    adminNote: String,

    createdAt: String,
    resolvedAt: String,
});

PojosMetadataMap.create<DisputeCreationRequest>('DisputeCreationRequest', {
    bookingId: String,
    damageTypeId: String,
    note: String,
    evidenceUrls: Array,
});

PojosMetadataMap.create<AdminReviewDisputeRequest>('AdminReviewDisputeRequest', {
    disputeId: String,
    isApproved: Boolean,
    adminNote: String,
});


// ========================================================
// 1) ResponseDTO → Model
// ========================================================
createMap<DisputeResponse, Dispute>(
    disputeMapper,
    'DisputeResponse',
    'Dispute'
);


// ========================================================
// 2) Model → CreationRequest
// (FE tạo khiếu nại)
// ========================================================
createMap<Dispute, DisputeCreationRequest>(
    disputeMapper,
    'Dispute',
    'DisputeCreationRequest'
);


// ========================================================
// 3) Model → AdminReviewRequest
// (admin xử lý tranh chấp)
// ========================================================
createMap<Dispute, AdminReviewDisputeRequest>(
    disputeMapper,
    'Dispute',
    'AdminReviewDisputeRequest'
);


// ========================================================
// disputeMap API – simple to use
// ========================================================
export const disputeMap = {
    /** Backend → FE model */
    fromResponse(dto: DisputeResponse): Dispute {
        return disputeMapper.map(dto, 'DisputeResponse', 'Dispute');
    },

    /** FE model → Create Dispute Request */
    toCreationRequest(model: Dispute): DisputeCreationRequest {
        return disputeMapper.map(model, 'Dispute', 'DisputeCreationRequest');
    },

    /** FE model → Admin review request */
    toAdminReviewRequest(model: Partial<Dispute> & { id: string; isApproved: boolean }): AdminReviewDisputeRequest {
        return {
            disputeId: model.id,
            isApproved: model.isApproved,
            adminNote: model.adminNote ?? ''
        };
    }
};
